### PR TITLE
mbedtls: add mbedtls port: v2.28.0

### PR DESCRIPTION
**Pull request by damianloew**
## Description

- mbedtls: add mbedtls port: v2.28.0

## Motivation and Context

Needed for azure iot sdk incoming port.

DONE: PD-272

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (running mbed tls tests on ia32-generic target - all passed excluding cases mentioned in README.md, tested in azure iot sdk c on armv7m7-imxrt106x).

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

https://github.com/phoenix-rtos/phoenix-rtos-doc/pull/119
## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: mbedtls/patches/04_set_devrandom_entropy.patch [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
++        ret = fread( p, 1, left, file );
++        if( ret == 0 && ferror( file ) )
++        {
++            fclose( file );
++            return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
++        }
++
++        p += ret;
++        left -= ret;
++        sleep( 1 );
```
Comment (MaciejPurski) [inline mbedtls/patches/04_set_devrandom_entropy.patch:106]: Is it necessary to sleep after every fread(), even if we get `len` bytes after the first call?

File: mbedtls/patches/03_rename_debug_var.patch [hunk 2]
```diff
@@ -0,0 +1,40 @@
+diff -ur '--exclude=*.git' mbedtls-2.28.0/programs/ssl/ssl_context_info.c mbedtls-2.28.0-phoenix/programs/ssl/ssl_context_info.c
+--- mbedtls-2.28.0/programs/ssl/ssl_context_info.c	2022-05-12 15:49:04.327065471 +0200
++++ mbedtls-2.28.0-phoenix/programs/ssl/ssl_context_info.c	2022-05-12 15:44:21.216459932 +0200
+@@ -118,7 +118,8 @@
+ FILE *b64_file = NULL;                  /* file with base64 codes to deserialize */
+ char conf_keep_peer_certificate = 1;    /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE from mbedTLS configuration */
+ char conf_dtls_proto = 1;               /* MBEDTLS_SSL_PROTO_DTLS from mbedTLS configuration */
+-char debug = 0;                         /* flag for debug messages */
++/* 'debug' has been already defined as global variable in Phoenix-RTOS */
```
Comment (MaciejPurski) [inline mbedtls/patches/03_rename_debug_var.patch:9]: `debug()` is a syscall, not a global variable.

I don't know, why this name conflict happens to be honest. mbedtls does not include `<sys/debug.h>`. Verify if this is really necessary.

**Review by MaciejPurski — CHANGES_REQUESTED** [review]

File: mbedtls/build.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	fi
+done
+
+# Convert ldflags to format recognizable by gcc, for example -q -> -Wl,-q
+LDFLAGS=$(echo " ${LDFLAGS}" | sed "s/\s/,/g" | sed "s/,-/ -Wl,-/g")
+
+# Flag that can be checked in makefiles
+export phoenix=1
+# Build and install tests if needed
+if [ "${LONG_TEST}" = "y" ]; then
```
Comment (nalajcie) [inline mbedtls/build.sh:41]: for readability reasons I would prefer to have the usual path first

File: mbedtls/build.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	(cd "${PREFIX_MBEDTLS_BUILD}/${MBEDTLS}" && make install all)
+
+	mkdir "${PREFIX_FS}/root/mbedtls_test_configs/"
+
+	for file in "${PREFIX_MBEDTLS_TESTS}"/*
```
Comment (nalajcie) [inline mbedtls/build.sh:46]: `for aaa in xxx; do`

**Review by nalajcie — CHANGES_REQUESTED** [review]
pointed out some issues, the PR might need a little more work before the merge :)

Did not try running it, just compiling.

File: mbedtls/patches/03_rename_debug_var.patch (hunk 2 above)
Comment (damianloew) [reply mbedtls/patches/03_rename_debug_var.patch:9]: Yeah , I will change it to syscall. I was looking for including something related to syscalls in mbedtls and the keyword `syscall` appears only in binaries of mbedtls programs. Couldn't find anything more about it

File: mbedtls/patches/01_set_destdir.patch [hunk 5]
```diff
@@ -0,0 +1,9 @@
+diff -ur '--exclude=*.git' mbedtls-2.28.0/Makefile mbedtls-2.28.0-phoenix/Makefile
+--- mbedtls-2.28.0/Makefile	2022-05-12 15:49:04.287065127 +0200
++++ mbedtls-2.28.0-phoenix/Makefile	2022-05-12 15:40:36.482048098 +0200
+@@ -1,4 +1,4 @@
+-DESTDIR=/usr/local
++DESTDIR=$(PREFIX_BUILD)
+ PREFIX=mbedtls_
```
[the remark this answers is not in the document]
Comment (damianloew) [reply mbedtls/patches/01_set_destdir.patch:7]: That's right I saw that destination directories are the same so left it, but didn't notice binaries. I agree on that the solution with additional `root` directory is better.

File: build.sh [hunk 6]
```diff
@@ -67,4 +67,6 @@ fi
 
 [ "${PORTS_OPENIKED}" = "y" ] && ./phoenix-rtos-ports/openiked/build.sh
 
+[ "${PORTS_MBEDTLS}" = "y" ] && ./phoenix-rtos-ports/mbedtls/build.sh
```
[the remark this answers is not in the document]
Comment (damianloew) [reply build.sh:70]: I will add it as a first port then. Similar situation is when using `azure sdk` with `mbedtls`

File: mbedtls/build.sh (hunk 3 above)
Comment (damianloew) [reply mbedtls/build.sh:41]: Good idea, I also think it's more readable

File: mbedtls/patches/03_rename_debug_var.patch (hunk 2 above)
Comment (nalajcie) [reply mbedtls/patches/03_rename_debug_var.patch]: I'm guessing but this is non-static global variable so has **external** linkage by default. This results in _multiple definitions of symbol `debug`_ in linking stage.

better solution would be to prepend the variable declaration with `static` keyword if that's the case

File: mbedtls/patches/04_set_devrandom_entropy.patch [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
++
++    ret = fread( p, 1, len, file );
++    if( ret == 0 && ferror( file ) ) 
++    {
++        fclose( file );
++        return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
++    }
++
++    fclose( file );
++    *olen = len;
```
Comment (nalajcie) [inline mbedtls/patches/04_set_devrandom_entropy.patch:103]: invalid, use `ret` here!

(not sure if our `/dev/random` implementation would ever return less than required number of bytes but it's more correct that way)

File: mbedtls/patches/04_set_devrandom_entropy.patch [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ /**
+  * \brief           Entropy poll callback for a non-volatile seed file
+diff -ur '--exclude=*.git' mbedtls-2.28.0/library/entropy.c mbedtls-2.28.0-phoenix/library/entropy.c
+--- mbedtls-2.28.0/library/entropy.c	2022-06-03 16:51:07.223241958 +0200
++++ mbedtls-2.28.0-phoenix/library/entropy.c	2022-06-03 16:14:49.600992572 +0200
+@@ -61,7 +61,6 @@
+ {
+     ctx->source_count = 0;
+     memset( ctx->source, 0, sizeof( ctx->source ) );
+-
```
Comment (nalajcie) [inline mbedtls/patches/04_set_devrandom_entropy.patch:45]: unrelated change

File: mbedtls/patches/03_avoid_relinking [hunk 9]
```diff
@@ -0,0 +1,12 @@
+diff -ur '--exclude=*.git' mbedtls-2.28.0/Makefile mbedtls-2.28.0-phoenix/Makefile
+--- mbedtls-2.28.0/Makefile	2022-06-03 14:45:36.268697648 +0200
++++ mbedtls-2.28.0-phoenix/Makefile	2022-06-03 14:55:26.338969692 +0200
+@@ -39,7 +39,7 @@
+ 	    if [ -x $$p ] && [ ! -d $$p ] ;     \
+ 	    then                                \
+ 	        f=$(PREFIX)`basename $$p` ;     \
+-	        cp $$p $(DESTDIR)/bin/$$f ;     \
```
Comment (nalajcie) [inline mbedtls/patches/03_avoid_relinking:8]: no, I was talking about installing **libraries** (it's already done correctly for **headers**). Maybe I was too brief in my explanation:
 - currently mbedtls will **always** update installed library files modification time - even if the library itself was not rebuilt (check this by comparing timestamps of `library/libmbedtls.a` with the installed one
 - `make` - when compiling a project:
   - creates dependency graph (which files depend on other files - for example  `foo.o` depends on `foo.c` and `foo.bin` depends on `foo.o` and `bar.o`)
   - `make` decides which parts of the dependency graph need to be rebuilt based on file modification times (if the target is older than the source - rebuild!)
   - as binaries linked against mbedtls will depend on `libmbedtls.a` - and it's timestamp updates with each compilation - they will always be (unnecessarily) re-linked

File: mbedtls/build.sh [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+		fi
+	done
+	# Files required for some tests
+	cp -r "${PREFIX_MBEDTLS_SRC}/tests/data_files" "${PREFIX_FS}/root/"
+fi
+
+# Copy built libraries and mbedtls header files to `lib` and `include` dirs
+# Note: assuming that files are present if `include/mbedtls` exists
+if ! [ -d "${PREFIX_BUILD}/include/mbedtls" ]; then
+	cp -r "${PREFIX_MBEDTLS_DESTDIR}/lib"/* "${PREFIX_BUILD}/lib"
```
Comment (nalajcie) [inline mbedtls/build.sh:69]: `-p` or `-a`

Use `PREFIX_H` and `PREFIX_A` here

File: mbedtls/build.sh [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	fi
+done
+
+# Convert ldflags to format recognizable by gcc, for example -q -> -Wl,-q
+LDFLAGS=$(echo " ${LDFLAGS}" | sed "s/\s/,/g" | sed "s/,-/ -Wl,-/g")
+
+# Flag that can be checked in makefiles
+export phoenix=1
+
+# Build and install tests if needed
```
Comment (nalajcie) [inline mbedtls/build.sh:42]: invalid comment :p

File: mbedtls/patches/03_avoid_relinking (hunk 9 above)
Comment (nalajcie) [reply mbedtls/patches/03_avoid_relinking:8]: alternative approach would be to not use `install` target at all but just to copy the needed files by hand

File: mbedtls/patches/04_set_devrandom_entropy.patch (hunk 7 above)
Comment (damianloew) [reply mbedtls/patches/04_set_devrandom_entropy.patch:103]: Right, thanks

File: mbedtls/patches/03_avoid_relinking (hunk 9 above)
Comment (damianloew) [reply mbedtls/patches/03_avoid_relinking]: Thanks for explaining that, firstly I didn't notice that there is `P` not `p`. I hope that `RPp` arguments will be ok then

File: mbedtls/patches/03_rename_debug_var.patch (hunk 2 above)
Comment (damianloew) [reply mbedtls/patches/03_rename_debug_var.patch]: `static` solves the problem, I agree on that it's a better solution

File: mbedtls/patches/03_avoid_relinking.patch [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+--- mbedtls-2.28.0/Makefile	2022-06-06 13:30:25.889607491 +0200
++++ mbedtls-2.28.0-phoenix/Makefile	2022-06-06 13:29:44.478798560 +0200
+@@ -30,9 +30,9 @@
+ 	cp -rp include/psa $(DESTDIR)/include
+ 
+ 	mkdir -p $(DESTDIR)/lib
+-	cp -RP library/libmbedtls.*    $(DESTDIR)/lib
+-	cp -RP library/libmbedx509.*   $(DESTDIR)/lib
+-	cp -RP library/libmbedcrypto.* $(DESTDIR)/lib
++	cp -RPp library/libmbedtls.*    $(DESTDIR)/lib
```
Comment (nalajcie) [inline mbedtls/patches/03_avoid_relinking.patch:11]: you can use `cp -a` instead (the same functionality)

File: mbedtls/build.sh [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+				b_install "${PREFIX_MBEDTLS_TESTS}/${test_executable}" /bin/
+			fi
+		fi
+	done
+	# Files required for some tests
+	cp -r "${PREFIX_MBEDTLS_SRC}/tests/data_files" "${PREFIX_FS}/root/"
+fi
+
+# Copy built libraries and mbedtls header files to `lib` and `include` dirs
+cp -rp "${PREFIX_MBEDTLS_DESTDIR}/lib"/* "${PREFIX_A}"
```
Comment (nalajcie) [inline mbedtls/build.sh:68]: you can use `cp -a` instead (the same functionality)

**Review by nalajcie — COMMENTED** [review]
nitpicking, otherwise LGTM

File: mbedtls/build.sh (hunk 13 above)
Comment (damianloew) [reply mbedtls/build.sh]: Thanks, I changed `-RP` and `-rp` to `-a` and squashed last commits into one. I also launched all mbedtls tests after rebasing and results are the same.

File: mbedtls/README.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    mbedtls_aes_context ctx;
+
+    mbedtls_aes_init( &ctx );
+    printf("Hello World!!\n");
+  
+    return 0;
+  }
+  ```
+
+- Note: Please remember that `PORTS_MBEDTLS` should be set to `y` in the specific building script in `_targets`/`_projects` directory or using an environment variable.
```
Comment (nalajcie) [inline mbedtls/README.md:51]: not necessarily in `_targets` dir

**Review by nalajcie — APPROVED** [review]
Looks good, maybe we should enable building this lib for one of the reference projects to test it from time to time by CI?

**Review by MaciejPurski — APPROVED** [review]
